src/assembler.cpp

```diff
@@ -303,30 +304,66 @@ void Assembler::LHU(GPR rd, int32_t imm, GPR rs) noexcept {
 }
 
 void Assembler::LI(GPR rd, uint32_t imm) noexcept {
-    const auto lower = imm & 0xFFF;
-    const auto upper = (imm & 0xFFFFF000) >> 12;
-    const auto simm = static_cast<int32_t>(imm);
-
-    // If the immediate can fit within 12 bits, we only need to emit an ADDI.
-    if (IsValidSigned12BitImm(simm)) {
-        ADDI(rd, x0, static_cast<int32_t>(lower));
-    } else {
-        const bool needs_increment = (lower & 0x800) != 0;
-        const auto upper_imm = needs_increment ? upper + 1 : upper;
-
-        // Note that we add 1 to the upper portion of the immediate if the lower
-        // immediate's most significant bit is set. This is necessary, as ADDI
-        // sign-extends its 12-bit immediate before performing addition.
-        //
-        // In the event of the sign-extension, this means that we'll be adding
-        // an equivalent of "lower - 4096" to the upper immediate.
-        //
-        // We add 1 to the upper part of the immediate. the upper part's least
-        // significant bit is bit 12. Adding 1 to this bit is equivalent to adding
-        // 4096, which counteracts the sign-extension, preserving the value.
+    // Depending on imm, the following instructions are emitted.
+    // hi20 == 0              -> ADDI
+    // lo12 == 0 && hi20 != 0 -> LUI
+    // otherwise              -> LUI+ADDI
+
+    /* Add 0x800 to cancel out the signed extension of ADDI. */
+    const auto hi20 = (imm + 0x800) >> 12 & 0xfffff;
+    const int32_t lo12 = static_cast<int32_t>(imm) & 0xfff;
+    GPR rs1 = zero;
+
+    if (hi20 != 0) {
+        LUI(rd, hi20);
+        rs1 = rd;
+    }
+
+    if (lo12 != 0 || hi20 == 0) {
+        ADDI(rd, rs1, lo12);
+    }
+}
+
+void Assembler::LI64(GPR rd, uint64_t imm) noexcept {
```

Comment:
Perhaps this should just be joined into LI so there's a single API to load immediates, rather than needing to flip-flop between the two depending on what you're trying to load (we haven't yet hit 1.0, so API breaks are allowable).

Reply:
Thanks for the review.

The difference between LI and LI64 is that the former can only be used in rv32 and the latter is only for rv64 -- they use ADDI and ADDIW for the 32bit part respectively. So I'm not sure how we can join this, do you have any idea in mind?

Reply:
Maybe we can add an XLEN (RV32/RV64/RV128) field to the Assembler? By that things like SLLI / SLLI64 can also be joined.

Reply:
Wouldn't it be sufficient to rename LI64 to LI, creating an overload specifically for 64-bit values?

Reply:
Actually, never mind, that wouldn't work, because it'd create ambiguous conversion scenarios

Reply:
Missed the XLEN comment because GitHub, but yeah, that would be a good idea. I'll add that to my TODO list.